### PR TITLE
[TorchTrain] Update train_state.current_loss to be the mean across all ranks

### PR DESCRIPTION
**Pull request by wz337**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #140
* __->__ #139
* #138

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -264,23 +264,18 @@ def main(job_config: JobConfig):
             if torch_profiler:
                 torch_profiler.step()
 
-            train_state.current_loss = loss.item()
+
+            avg_loss = dist_mean(loss.item(), world_mesh)
```
Comment (tianyu-l) [inline train.py:268]: May I ask why we need to do all-reduce among all ranks, as the TP ranks should share the same loss?

A more general question as I'm lacking context on this: why we need to collect all-reduced loss every single iteration?

Thanks!

Comment (wz337) [reply train.py:268]: This is more for checkpoints. Since before checkpoint, we need the loss to be the same across all ranks, as they share the same FQN. If they are different, then we are just saving the data from one rank and broadcasting that when loading back. 

Seems Mosaic, XLFormers, and Ads all do similar things like this.

Comment (awgu) [reply train.py:268]: Sorry @wz337. The xlformers logging is a bit more complicated than I thought. It might be worth a closer look. I am not sure if the logic to all-reduce losses runs by default since it is only if the `loss_tracker is not None`.

Comment (tianyu-l) [reply train.py:268]: That makes sense to me. But still, IIRC, you may need to change `world_mesh` to `dp_mesh`; o/w the collectives would fail when 2D parallelism is enabled.

Comment (wz337) [reply train.py:268]: > Sorry @wz337. The xlformers logging is a bit more complicated than I thought. It might be worth a closer look. I am not sure if the logic to all-reduce losses runs by default since it is only if the `loss_tracker is not None`.

Hmm. I can also potentially do a all_reduce for losses instead of current_loss when checkpoint, but it would be a little bit messier cuz the losses of train_state all_reduce would need to happen inside checkpointer.py, which should be unaware of the TrainState.

Comment (wz337) [reply train.py:268]: > That makes sense to me. But still, IIRC, you may need to change `world_mesh` to `dp_mesh`; o/w the collectives would fail when 2D parallelism is enabled.

Ye. Makes sense!
